### ISSUE TITLE
Bug in NDFD time intervals when interval units different from pds units

### ISSUE THREAD

**Issue by JohnLCaron** [Unidata/netcdf-java#606]
Ticket ID: JMQ-641023

Some NDFD datasets use time intervals that are different from PDS time unit. This is not being handled correctly in version 5 and 6. Haven't tested 4.6, but maybe broken there also.

Im working on a fix for 6, probably can be easily backported to 5.

In looking at this, Im reconsidering how time intervals are calculated, and its possible they are still wrong, at least for NDFD.
Spot checking GFS/global 0.25, NAM conus 12, NAM firewxnest, RR conus 13, CMC/RDPS/NA_15km appear ok. NOAA_GSD/HRRR/CONUS_3km/surface has 88/122261 possibly wrong. 

1. How to get the truth? Im guessing we need to fire up the "standard" NCEP library and compare. Its possible that different groups handle it differently, since I was following someone's advice. Or that the only bug is when the units change.
2. Check to see what 4.6 is doing.
3. Fix 6 and then 5 once we know what its supposed to be.
4. What about ECMWF and others?

**Comment by JohnLCaron** [comment]
[Uploading ds.pop12-18.zip…]()

I am attaching a file with 5 NDFD GRIB2 records from Nov 2020, of NDFD parameter 0-1-8 (Total precipitation).
The first record is problematic for us.

The reference date=2020-11-18T17:30:00Z
The forecast time =0 in units of minutes
The time interval length is 12 in units of hours.

On the face of it, this gives a forecast interval of

  TimeIntervalNew=(2020-11-18T17:30:00Z,2020-11-19T05:30:00Z)
 
So then, what is the meaning of octets 48-54 "end of overall time interval" showing TimeIntervalEnd=2020-11-19T00:00:00Z
Our previous algorithm calculates the beginning of the interval as  TimeIntervalEnd minus the time interval length giving:

   TimeInterval   =(2020-11-18T12:00:00Z,2020-11-19T00:00:00Z)
   
For all of the other records, the two methods agree.
So, either "forecast time =0" is wrong or TimeIntervalEnd=2020-11-19T00:00:00Z.

Can you shed any light on this?

Here are the relevant parts of that first record, from our decoder:  
````
Grib2IndicatorSection
 Discipline = (0) Meteorological products
 Length     = 360491

Grib2IdentificationSection
 Center        = (8) US National Weather Service Telecommunications Gateway (NWSTG)
 SubCenter     = (-9999) null
 Master Table  = 1
 Local Table   = 0
 RefTimeSignif = 1 (Start of forecast)
 RefTime       = 2020-11-18T17:30:00Z
 RefTime Fields = 2020-11-18 17:30:0
 ProductionStatus      = 0 (Operational products)
 TypeOfProcessedData   = 1 (Forecast products)

Grib2ProductDefinitionSection
 Interval     = (2020-11-18T12:00:00Z,2020-11-19T00:00:00Z)

(4.9) Product definition template 4.9 - probability forecasts at a horizontal level or in a horizontal layer in a continuous or non-continuous time interval
  1:                                                                                 PDS length == 71
  5:                                                                                    Section == 4
  6:                                                Number of coordinates values after Template == 0
  8:                                                         Product Definition Template Number == 9
 10:                                                                         Parameter category == 1
 11:                                                                           Parameter number == 8
 12:                                                                 Type of generating process == 2 (table 4.3: Forecast)
 13:                   Background generating process identifier (defined by originating centre) == 0 (table ProcessId: Table ProcessId code 0 not found)
 14:                     Forecast generating process identifier (defined by originating centre) == 0 (table ProcessId: Table ProcessId code 0 not found)
 15:                                                 Hours after reference time of data cut-off == 255
 17:                                               Minutes after reference time of data cut-off == 255
 18:                                                            Indicator of unit of time range == 0 (table 4.4: Minute)
 19:                                                 Forecast time in units defined by octet 18 == 0
 23:                                                                Type of first fixed surface == 1 (table 4.5: Ground or water surface)
 24:                                                        Scale factor of first fixed surface == 0
 25:                                                        Scaled value of first fixed surface == 0
 29:                                                               Type of second fixed surface == 255 (table 4.5: Missing)
 30:                                                       Scale factor of second fixed surface == 129
 31:                                                       Scaled value of second fixed surface == -9999
 35:                                                                Forecast probability number == 255
 36:                                                     Total number of forecast probabilities == 255
 37:                                                                           Probability type == 1 (table 4.9: Probability of event above upper limit)
 38:                                                                Scale factor of lower limit == 129
 39:                                                                Scaled value of lower limit == -9999
 43:                                                                Scale factor of upper limit == 3
 44:                                                                Scaled value of upper limit == 254
 48:                                                       Year of end of overall time interval == 2020
 50:                                                      Month of end of overall time interval == 11
 51:                                                        Day of end of overall time interval == 19
 52:                                                       Hour of end of overall time interval == 0
 53:                                                     Minute of end of overall time interval == 0
 54:                                                     Second of end of overall time interval == 0
 55: n - number of time range specifications describing the time intervals used to calculate the statistically processed field == 1
 56:                             Total number of data values missing in the statistical process == 0
 60: Statistical process used to calculate the processed field from the field at each time increment during the time range == 1 (table 4.10: Accumulation)
 61:        Type of time increment between successive fields used in the statistical processing == 255 (table 4.11: Missing)
 62:         Indicator of unit of time for time range over which statistical processing is done == 1 (table 4.4: Hour)
 63: Length of the time range over which statistical processing is done, in units defined by the previous octet == 12
 67:             Indicator of unit of time for the increment between the successive fields used == 1 (table 4.4: Hour)
 68:           Time increment between successive fields, in units defined by the previous octet == 0
 72:                                      As octets 60 to 71, next innermost step of processing == -9999
 84: Additional time range specifications, included in accordance with the value of n. Contents as octets 60 to 71, repeated as necessary. == -9999
````

**Comment by JohnLCaron** [comment]
From Timothy Boyer:

Strictly speaking, the forecast time = 0 is incorrect (and the end
forecast time IS correct).  The begin time for the 12 hour interval
ending at 2020-11-19 00Z is 2020-11-18 12Z, which would make the correct
forecast hour = -5.5 hours.  However, our encoding software will not
allow us to specify negative values for the forecast hour.  Therefore,
we've adopted the convention of assigning the forecast hour = 0 whenever
the forecast reference time falls between the start and end of the
forecast time range (12 hours in this case).   For these cases, using
the reference time and forecast time to compute the beginning of the
forecast interval will obviously give you an incorrect value.

The way to compute the beginning of the forecast time range for this
case is to subtract the statistical time range (12 hours) from the end
time of the forecast range (2020-11-19 00Z) which will give you the
correct time of 2020-11-18 12Z.

The NDFD elements where you'll run into this issue are ones with Product
Definition Templates = 8, 9, or 10.
